custom_components/amplipi/media_player.py

```diff
@@ -566,24 +576,49 @@ def __init__(self, namespace: str, zone, group,
         self._available = False
         self._extra_attributes = []
         self._attr_device_class = MediaPlayerDeviceClass.SPEAKER
+        self._is_off = False
+
+    async def async_toggle(self):
+        if self._is_off:
+            await self.async_turn_on()
+        else:
+            await self.async_turn_off()
+
+    async def async_turn_on(self):
+        if self._is_group:
+            _LOGGER.info(f"Turning group {self._name} on")
+            await self._update_group(
+                MultiZoneUpdate(
+                    groups=[self._group.id],
+                    update=ZoneUpdate(
+                        source_id=-1,
+                    )
+                )
+            )
+        else:
+            _LOGGER.info(f"Turning zone {self._name} on")
+            await self._update_zone(ZoneUpdate(
+                source_id=-1,
+            ))
+        self._is_off = False
 
     async def async_turn_off(self):
-        if self._current_source is not None:
-            if self._is_group:
-                _LOGGER.info(f"Disconnecting zones from source {self._current_source}")
-                await self._update_group(
-                    MultiZoneUpdate(
-                        groups=[self._group.id],
-                        update=ZoneUpdate(
-                            source_id=-1,
-                        )
+        if self._is_group:
+            _LOGGER.info(f"Turning group {self._name} off")
+            await self._update_group(
+                MultiZoneUpdate(
+                    groups=[self._group.id],
+                    update=ZoneUpdate(
+                        source_id=-2,
```

Comment:
```suggestion
        # update zone/group to have a disconnected source state that indicates to HA that the zone/group is off
        source_off_update = ZoneUpdate(source_id=-2)
        if self._is_group:
            _LOGGER.info(f"Turning group {self._name} off")
            await self._update_group(
                MultiZoneUpdate(
                    groups=[self._group.id],
                    update=source_off_update
```

Reply:
moving the underlying update to a common place gives a single place to document what we are doing